src/tir/transforms/hoist_if_then_else.cc

```diff
@@ -0,0 +1,374 @@
+    // Check if the loop position is higher than the parent loop position
+    for (auto var : if_var_list_) {
+      if (GetParentLoopPos(var_for_map_[var]) >= loop_pos) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  int GetParentLoopPos(const Object* node) {
+    for (size_t i = 0; i < ordered_for_list_.size(); ++i) {
+      if (ordered_for_list_[i] == node) {
+        return i;
+      }
+    }
+    return -1;
+  }
+
+  void InitRecorder() { hoist_for_if_recorder = std::make_tuple(false, nullptr, nullptr); }
+
+  void StopRecording() { is_recorder_on = false; }
+
+  bool IsRecordingOn() { return is_recorder_on; }
+
+  void StartOrAddRecord(const ForNode* op) {
+    is_recorder_on = true;
+    if (!var_for_map_.count(op->loop_var.get())) {
+      var_for_map_.insert({op->loop_var.get(), op});
+    }
+    ordered_for_list_.emplace_back(op);
+  }
+
+  void RemoveRecord(const ForNode* op) {
+    StopRecording();
+    var_for_map_.erase(op->loop_var.get());
+    if (ordered_for_list_.size() > 0) ordered_for_list_.pop_back();
+  }
+
+  void StopAndAddRecord(const ForNode* for_node, const IfThenElseNode* if_node) {
+    hoist_for_if_recorder = std::make_tuple(true, for_node, if_node);
+    StopRecording();
+  }
+
+  void UpdateAttrVarList(const AttrStmtNode* op) {
+    if (const auto* iv = op->node.as<IterVarNode>()) {
+      attr_var_list_.insert(iv->var.get());
+    } else if (const auto* iv = op->node.as<VarNode>()) {
+      attr_var_list_.insert(iv);
+    }
+  }
+
+  void RemoveAttrVarList(const AttrStmtNode* op) {
+    if (const auto* iv = op->node.as<IterVarNode>()) {
+      attr_var_list_.erase(iv->var.get());
+    } else if (const auto* iv = op->node.as<VarNode>()) {
+      attr_var_list_.erase(iv);
+    }
+  }
+
+  bool CheckAttrVar() {
+    for (auto var : if_var_list_) {
+      if (attr_var_list_.count(var)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  std::vector<const ForNode*> ordered_for_list_;
+
+  std::vector<const VarNode*> if_var_list_;
+
+  std::unordered_set<const VarNode*> attr_var_list_;
+
+  VarForMap var_for_map_;
+
+  bool is_if_cond{false};
+  bool is_recorder_on{false};
+};
+
+class IfThenElseHoister : public StmtMutator {
+ public:
+  IfThenElseHoister() : hoist_selector(HoistCandidateSelector()) {}
+
+  Stmt VisitAndMutate(Stmt stmt) {
+    hoist_selector(stmt);
+    Stmt stmt_copy = std::move(stmt);
+
+    while (hoist_selector.RecordingComplete()) {
+      target_for = hoist_selector.GetTargetForNode();
+      target_if = hoist_selector.GetTargetIfNode();
+
+      stmt_copy = operator()(stmt_copy);
+
+      hoist_selector.ResetRecorder();
+      hoist_selector(stmt_copy);
+    }
+
+    // Support SSA Form
+    stmt_copy = ConvertSSA(stmt_copy);
+    return stmt_copy;
+  }
+
+  Stmt VisitStmt_(const ForNode* op) final {
+    if ((!is_updating) && (target_for == op)) {
+      is_updating = true;
+      is_then_case = true;
+      Stmt then_case = StmtMutator::VisitStmt_(op);
```

Comment:
`VisitStmt(op->then_case)`

Reply:
Sorry, i could not get your point clearly here. Here the op node is `ForNode`, it does not have `then_case`.
Would you please help me understand better!

Reply:
My mistake.

Reply:
No prob :) 
Closing this comment and the below one. Thanks!